Sebastian.p/orth dgp

SebastianPopescu opened the pull request:
Implementation of ["Sparse Orthogonal Variational Inference for Gaussian Processes"](https://arxiv.org/abs/1910.10596v3).

Created the following folders:
-- conditionals: needs a specialized form to account for the two different GPs that have to be summed g() and h() as in the paper.
--covariances: needed to compute Cvv and Cvf as in the paper. These covariances rely on the other set of inducing points.
--posteriors.py: different conditionals are needed here.
-- conditionals.util.conditional_GP_maths might have to be re-designed or at least have its name changed to something more sensible.

SebastianPopescu committed:
implementation of DistDGPs

SebastianPopescu committed:
implementation of orthogonal sparse GPs

SebastianPopescu committed:
implementation of Orthogonal Sparse GPs

SebastianPopescu committed:
changes in requirements

SebastianPopescu committed:
minor updated

SebastianPopescu committed:
latest_update

SebastianPopescu committed:
updated codebase

SebastianPopescu committed:
added tests

SebastianPopescu committed:
full covariance function coverage

hstojic committed:
removing files from git

hstojic committed:
format

hstojic commented:
cleaned up a bit, mypy still complains on quite a few things, but content wise good enough for a review

avullo commented on docs/my_notebooks/motor_het_student_orth_dgp.py:
```diff
@@ -0,0 +1,307 @@
+    def motorcycle_data():
+        """ Return inputs and outputs for the motorcycle dataset. We normalise the outputs. """
+        import pandas as pd
+        df = pd.read_csv("/home/sebastian.popescu/Desktop/my_code/GP_package/docs/notebooks/data/motor.csv", index_col=0)
```
This should be accessible outside local filesystem.

avullo commented on gpflux/models/deep_gp.py:
```diff
@@ -291,6 +311,12 @@ def as_prediction_model(
         return model_class(self.inputs, outputs)
 
 
+class OrthDeepGP(DeepGP):
+    """
+    Orthogonal Deep Gaussian Processes
+    """
+
+
```
This is also declared in the `gpflux.models.orth_deep_gp` module.

SebastianPopescu replied:
ahh right, forgot to delete that file.

avullo committed:
Address linting issues.

avullo commented on gpflux/covariances/multioutput/cvvs.py:
```diff
@@ -0,0 +1,185 @@
+    jittermat = tf.eye(inducing_variable_v.num_inducing, dtype=Cvv.dtype)[None, :, :] * jitter
+    return Cvv + jittermat
+
+    Kmms = [Kuu(f, k) for f, k in zip(inducing_variable.inducing_variable_list, kernel.kernels)]
```
@SebastianPopescu is this `inducing_variable_u` or `inducing_variable_v`?

SebastianPopescu replied:
uups..sorry..I again forgot to remove line 182

avullo commented on gpflux/covariances/multioutput/cvvs.py:
```diff
@@ -0,0 +1,185 @@
+
+    Kmms = [Kuu(f, k) for f, k in zip(inducing_variable.inducing_variable_list, kernel.kernels)]
+    Kmm = tf.stack(Kmms, axis=0)
+    jittermat = tf.eye(inducing_variable.num_inducing, dtype=Kmm.dtype)[None, :, :] * jitter
```
Same here.

SebastianPopescu replied:
yeah this part has to be removd..sorry

avullo committed:
Addressing some more lint errors.

SebastianPopescu committed:
fixed some bugs

avullo committed:
Mypy fixes.

avullo committed:
Merge branch 'sebastian.p/orth_dgp' of github.com:secondmind-labs/GPflux into sebastian.p/orth_dgp

avullo committed:
Black.

SebastianPopescu committed:
fixed some tests

SebastianPopescu committed:
fixed some tests

SebastianPopescu committed:
tests functioning

SebastianPopescu committed:
moved notebooks

avullo committed:
Moving to jupytext.

avullo committed:
Adding reference and contributors.

avullo committed:
Can reuse the GPflow het likelihood.

avullo committed:
Formatting.

sc336 commented:
The covariances, conditionals and posteriors probably ought to go in GPflow instead.  Maybe we can split these out into a separate PR?

sc336 committed:
Last minute correction by Sebastian during briefing

sc336 committed:
Sebastian thinks this is no longer needed

sc336 committed:
Format

sc336 committed:
format

avullo committed:
- Use configurations with models and likelihoods
- Switch to factories for constant input dim architecture in the normal and orthogonal case.

avullo committed:
Copyright and formatting.

avullo committed:
Merge remote-tracking branch 'origin/sebastian.p/orth_dgp' into sebastian.p/orth_dgp

# Conflicts:
#	docs/notebooks/deep_cde.py

avullo committed:
No longer needed.

avullo commented:
@hstojic @SebastianPopescu note how I've merged all the different model architecture builders. I've tried adding heteroskedastic likelihoods but at the moment some tests with this likelihood are failing, which I have explicitly marked as such.

avullo committed:
Validate model config and test.

avullo committed:
Parametrise tests.

avullo committed:
Removed redundant notebooks, finished main one.

avullo committed:
Mypy fixes.

avullo committed:
Typo.

avullo committed:
New config and architecture factory infrastructure.

avullo committed:
Formatting.

avullo committed:
API change.

avullo committed:
Must be test skipped.

avullo committed:
Remove unused components.

avullo commented:
@SebastianPopescu the PR is now failing on test coverage, mainly due to the untested GPflow customisations.
With a closer look, it seems there are components which have exact duplicates on GPflow, e.g. https://github.com/GPflow/GPflow/blob/develop/gpflow/kullback_leiblers.py

Why do we need to add them here?

avullo committed:
Merge branch 'develop' into sebastian.p/orth_dgp

avullo committed:
More type-safety.

avullo committed:
Remove gpflow duplicate.

avullo committed:
Merge branch 'develop' into sebastian.p/orth_dgp

avullo committed:
Fix assertion considering numpy array.

avullo committed:
Refactoring in an attempt to increase coverage.

avullo committed:
Both models are defined in the same file, so same test module.

avullo committed:
Unused module.

hstojic commented on docs/notebooks/plotting_functions.py:
```diff
@@ -0,0 +1,177 @@
+            label="Class 1",
+        )
+
+        # NOTE -- don't need this
```
we can remove the comments right?

hstojic commented on docs/notebooks/plotting_functions.py:
```diff
@@ -0,0 +1,177 @@
+import numpy as np
```
we had plotting functions like these in trieste but users complained about not being to able to easily run notebooks like that so we moved them to an [`experimental` package](https://github.com/secondmind-labs/trieste/tree/develop/trieste/experimental/plotting), from which they can be then imported

hstojic commented on docs/notebooks/deep_cde.ipynb:
```diff
@@ -2,18 +2,22 @@
  "cells": [
```
how come this notebook was changed? 
(probably a separate issue but we shouldn't be having ipynb files in the repo...)

avullo replied:
It was originally as a ipynb file, then I translated it to a jupytext script just like the rest, but then I've discovered why it was kept as a notebook. The tests take an enormous amount of time. I have then translated it back to a notebook.

hstojic commented on gpflux/architectures/factory.py:
```diff
@@ -0,0 +1,230 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""This module defines factories for various Deep GP architectures"""
```
great solution @avullo!

hstojic commented on gpflux/architectures/config.py:
```diff
@@ -0,0 +1,130 @@
+    """
+    The number of inducing variables, *M*. The Deep GP uses the same number
+    of inducing variables in each layer.
+    """
```
I guess the docstrings should differ here?

hstojic commented on gpflux/architectures/factory.py:
```diff
@@ -0,0 +1,230 @@
+def build_constant_input_dim_deep_gp(
+    model_config: ModelHyperParametersConfig, X: np.ndarray
+) -> DeepGP:
+    if X.dtype != default_float():
```
@avullo any significant changes in the bodies of the functions or just copy paste?

avullo replied:
No significant changes, just minor refactoring and type checking adjustments.

hstojic commented on gpflux/models/deep_gp.py:
```diff
@@ -291,6 +311,12 @@ def as_prediction_model(
         return model_class(self.inputs, outputs)
 
 
+class OrthDeepGP(DeepGP):
```
@avullo I'm a bit confused, do we really need this class at the end?

hstojic reviewed (commented):
provided some quick comments...
not sure what to do about testing those conditionals, posteriors etc

avullo commented:
> provided some quick comments... not sure what to do about testing those conditionals, posteriors etc

I'm waiting for what @sc336 has to say, if he manages to move all this stuff to GPflow in https://github.com/GPflow/GPflow/pull/2027

avullo committed:
Addresing PR's comment.

avullo committed:
Expand tests to increase coverage.

avullo committed:
Expand tests to increase coverage.

avullo committed:
Address PR's comment.

avullo committed:
Rolling back changes.

vdutor commented on gpflux/covariances/multioutput/cvvs.py:
```diff
@@ -0,0 +1,213 @@
+
+        lista.append(Cvv(ind_var_u, ind_var_v, k, L_Kuu=l_kuu))
+
+    """
+    lista =  [
+>>>>>>> 53dae0f597e5aa12eea4bc13839a3c94807acd36
+            Cvv(ind_var_u, ind_var_v, k, L_Kuu=l_kuu)
+            for ind_var_u, ind_var_v, l_kuu, k in zip(
+                inducing_variable_u.inducing_variable_list,
+                inducing_variable_v.inducing_variable_list,
+                L_Kuu,
+                kernel.kernels,
+            )
+        ]
+    """
```
What's the purpose of this block of text?

avullo replied:
I think it's a leftover from a previous attempt, can be removed.
